src/sync/coll.rs

```diff
@@ -153,7 +153,7 @@ where
         &self,
         pipeline: impl IntoIterator<Item = Document>,
         options: impl Into<Option<AggregateOptions>>,
-    ) -> Result<Cursor> {
+    ) -> Result<Cursor<Document>> {
```

Comment:
We could consider returning a `Cursor<O>` here (where `O` is a generic parameter to the `aggregate` method, not the `T` from the `Collection`). Or should we have a separate `aggregate_with_output_type` or something? I think with aggregation it's more likely that a user might just want to retrieve a document than with a `find`, since they may mutate the output along the way, so it's a little riskier to have this one be pure generic. Thoughts?

Reply:
per slack conversation: seems we're all on the same page the `T` from the collection should not be used since people often do transformations here.

as for whether to require a type be specified in this method (which I was not thinking about the option of in the slack discussion), or just having a separate `aggregate_with_type` I don't have particularly strong feelings. I agree with your comment that it does seem a bit at odds with the choice to make requiring a type the default for `db.collection`. OTOH maybe it's less likely that people will bother writing custom `Deserialize` types for the output types of single methods?

Reply:
Yeah, my original inclination was to just leave this as `Document` by default and introduce the new separate method, since I imagine it'll be pretty common to use `Document` at least while prototyping the pipeline. I'm starting to flip back to wanting to maintain consistency though, given how easy it is to just append a `::<Document>` to the end of the `aggregate` call. Especially in the face of complex pipelines, handling the generic type would be a piece of cake.

_However_

This got me trying it out and I noticed an interesting / frustrating technical challenge to doing it the `Cursor<O>` way: turbofish cannot be used in conjunction with `impl Trait` in the argument position (yet, see https://github.com/rust-lang/rust/issues/83701).

e.g.
```rust
fn foo<T>(bar: impl ToString) { ... }
foo::<usize>("blah"); // compile error!
```
yields the following compiler error:
```
error[E0632]: cannot provide explicit generic arguments when `impl Trait` is used in argument position
  --> src/main.rs:21:11
   |
21 |     foo::<usize>("hello");
   |           ^^^^^ explicit generic argument not allowed

```
So you'd have to do it the explicit type parameter way:
```rust
let x: Cursor<Document> = coll.aggregate(vec![], None).await?
```
Which could be pretty frustrating.

This will make it tough to implement the one that takes in a generic argument, since we use impl Trait in all our CRUD methods for our options types (e.g. to be able to say `options` instead of having to say `Some(options)`). This makes me think we may want to do the builder approach here and have the method be on `Cursor` actually, something like:

```rust
let cursor = coll.aggregate(vec![ ... ], None).await?.with_type::<MyDeserializeType>();
```
This is pretty ugly admittedly, but I think it's preferable to breaking consistency with how options are provided to the other CRUD methods or _requiring_ users to use explicit type annotations instead of turbofish.

Reply:
I agree changing the way we pass options either here or everywhere doesn't seem like a great path forward.

is turbofish common/used rather than type annotations that people would be running into this a lot?

adding the method to cursor seems reasonable though it does look clunky.... it also seems consistent with the `clone_with_type` behavior that we currently have on `Collection` to change the generic type. and maybe it would be useful in other use cases as well - e.g. if you provide a `projection` via `FindOptions` that changes the shape of your documents as well, and you could call this on the resulting cursor.

Reply:
Yeah none of these options is very ideal to me; `with_type` seems like it would be more ergonomic for chaining method calls together, since providing a type annotation would require breaking out the call to `aggregate` into a separate line. Updating our CRUD API to use the builder pattern would get rid of this issue, right?

In any case since we're opting to go the `aggregate_with_type` route, I think we can defer coming to a conclusion here til we implement that (and make a decision regarding the builder pattern).

Reply:
I think even with the builder pattern for options/sessions we'd still have the same issue unfortunately, since we use `impl IntoIterator` for the pipeline.

Also, to clarify, I don't think `aggregate_with_type` method will really be possible, since we'd have to ditch the `impl Trait` arguments, so I think we kind of have to do the `Cursor::with_type` approach. That being said, we don't need to implement that functionality now.

Reply:
Ah right, I forgot we use `impl` for the pipeline.